File: src/#1019/solution.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func nextLargerNOdes(head *ListNode) []int {
	if head == nil {
		return []int{}
	}

	nodes := []int{}
	for head != nil {
		nodes = append(nodes, head.Val)
		head = head.Next
	}

	larger := make([]int, len(nodes))
	stack := [][]int{}

	for i := 0; i < len(nodes); {
		for len(stack) != 0 && nodes[i] > stack[len(stack)-1][0] {
			top := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			larger[top[1]] = nodes[i]
		}
		stack = append(stack, []int{nodes[i], i})
		i++
	}

	for _, v := range stack {
		larger[v[1]] = 0
	}
	
	return larger
}

File: src/#1019/solution.js
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val
  this.next = next === undefined ? null : next
}

var nextLargerNodes = function (head) {
  if (head === null) {
    return []
  }
  let stack = []
  let res = []
  let i = 0
  let current = head
  while (current) {
    if (stack.length) {
      while (stack.length && current.val > stack[stack.length - 1].val) {
        res[stack.pop().index] = current.val
      }
    }
    stack.push({
      index: i,
      val: current.val,
    })
    i++
    current = current.next
  }
  stack.forEach(e => {
    res[e.index] = 0
  })
  return res
}
